returnn/datasets/postprocessing.py

```diff
@@ -187,14 +194,18 @@ def init_seq_order(
         assert self._dataset is not None
         self._dataset.init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
         self._data_iter = enumerate(self._build_mapping_iter())
+        self._data_iter_produced_num_seqs = 0
         self._seq_list_for_validation = seq_list
-        if self._map_seq_stream is None:
-            # If we don't have an iterable mapper we know the number of segments exactly
-            # equals the number of segments in the wrapped dataset
+        if self._map_seq_stream is None or self._map_seq_stream_preserves_num_seqs:
+            # If we don't have an iterable mapper (or the user explicitly specifies this),
+            # we know the number of segments exactly equals the number of segments in the wrapped dataset
             try:
                 self._num_seqs = self._dataset.num_seqs
             except NotImplementedError:
-                pass  # some datasets don't know their num_seqs
+                # some datasets don't know their num_seqs
+                assert (
+                    not self._map_seq_stream_preserves_num_seqs
+                ), "_map_seq_stream_preserves_num_seqs is True, but wrapped dataset does not know its num_seqs"
```

Comment:
I think `map_seq_stream_preserves_num_seqs` does not necessarily imply that the sub dataset knows the `num_seqs`, so I would not throw an error here.
```suggestion
                pass  # some datasets don't know their num_seqs
```

Reply:
I'm a bit torn because as a user, I would explicitly set this option because I want the dataset to set the property num_seqs and expect it to fail when it can't.
Ultimately I agree though, that based on the name of the attribute, this should not be an error

Reply:
I can imagine that in some Sisyphus recipe, you would automatically have some code snippet which adds some Laplace (or some other custom user post-proc func) and just always set this flag `map_seq_stream_preserves_num_seqs=True` because the user knows that this post-proc func preserves num seqs, but then it would fail now for datasets where it is not known, so having this check just makes this more complicated and less generic.